### ISSUE TITLE
AWS Amplify Auth Custom Flow Never Completes

### ISSUE THREAD

**Issue by kevin-agrology** [aws-amplify/amplify-ios#1686]
### Describe the bug

My iOS app is using Amplify Auth to sign in to an existing User Pool. I followed the Amplify doc for [Sign in with custom flow][1].

    func signIn(username: String) {
        Amplify.Auth.signIn(username: username, password: "dummyPassword") { result in
            switch result {
            case .success(let authResult):
                if case .confirmSignInWithCustomChallenge = authResult.nextStep {
                    // Ask the user to enter the custom challenge.
                } else {
                    print("--> Sign in succeeded")
                }
            case .failure(let error):
                print("--> Sign in failed \(error)")
            }
        }
    }

The `signIn()` call works as expected, and a verification email is received. The OTP (aka challenge response) is used to call `confirmSignIn()`:

```
func customChallenge(response: String) {
        Amplify.Auth.confirmSignIn(challengeResponse: response) { result in
            switch result {
            case .success(let authResult):
                print("--> Confirm sign in succeeded. isSignedIn?: \(authResult.isSignedIn)")
            case .failure(let error):
                print("--> Confirm sign in failed \(error)")
            }
        }
    }
```

which also succeeds, but the user is not signed in because `authResult.isSignedIn = false`.

We also have a web version of this auth flow, but it succeeds in signing in the user.

  [1]: https://docs.amplify.aws/lib/auth/signin_with_custom_flow/q/platform/ios/

Note I am not using Amplify CLI because I need to use an existing Cognito User Pool, and so I've only edited `amplifyconfiguration.json` and `awsconfiguration.json` to use the existing User Pool PoolId, AppClientId and Region.

I've been following the Amplify docs very closely. This should be work, right?

### Steps To Reproduce

```swift
Steps to reproduce the behavior:
1. Setup iOS app to target an existing User Pool via awsconfigure.json. This application should use custom auth for a passwordless flow.
2. Call Amplify.Auth.signIn() for a user that exists/is registered in the User Pool.
3. On success, call Amplify.Auth.confirmSignIn() with the challenge response
4. See that the call succeeds, but the user remains nil
```

### Expected behavior

The user is signed in upon successful challenge response API call.

### Amplify Framework Version

7.6.24

### Amplify Categories

Auth

### Dependency manager

Swift PM

### Swift version

4.2

### CLI version

7.6.24

### Xcode version

13.3

### Is this a regression?

Yes

### Device

iPhone 11

### iOS Version

15.4

### Specific to simulators

No

**Comment by royjit** [comment]
@kevin-agrology Does the `authResult.nextStep` has some value? 

Also, can you share the lambda trigger logic here? Please note that Define Auth lambda trigger should bypass the SRP challenge - https://docs.amplify.aws/lib/auth/signin_with_custom_flow/q/platform/ios/#lambda-trigger-setup

**Comment by kevin-agrology** [comment]
> @kevin-agrology Does the `authResult.nextStep` has some value?
> 
> Also, can you share the lambda trigger logic here? Please note that Define Auth lambda trigger should bypass the SRP challenge - https://docs.amplify.aws/lib/auth/signin_with_custom_flow/q/platform/ios/#lambda-trigger-setup

authResult.nextStep = `confirmSignInWithCustomChallenge(Optional(["email": "<my registered email>"]))`

Our Define Auth trigger should be the same, but in Go: 

```
package main

import (
	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
)

func main() {
	lambda.Start(lambdaHandler)
}

func lambdaHandler(e events.CognitoEventUserPoolsDefineAuthChallenge) (events.CognitoEventUserPoolsDefineAuthChallenge, error) {

	response := events.CognitoEventUserPoolsDefineAuthChallengeResponse{ChallengeName: "CUSTOM_CHALLENGE"}
	i := -1
	l := len(e.Request.Session)
	authFailed := func() {
		response.IssueTokens = false
		response.FailAuthentication = true
	}

	if l == 0 {
		// if there are no session callenge results yet, set it up for the user
		goto respond
	}

	// from this point, we know that l is > 0 when reaching into the slice

	for j, sess := range e.Request.Session {
		if sess.ChallengeName == "SRP_A" {
			sess.ChallengeName = "CUSTOM_CHALLENGE"
			goto respond
		}
		if sess.ChallengeName == "CUSTOM_CHALLENGE" {
			i = j
			break
		}
	}
	if i == -1 {
		goto respond
	}

	if l > 3 || (l == 3 && !e.Request.Session[l-1].ChallengeResult) {
		// The user provided a wrong answer 3 times; fail auth
		authFailed()
		goto respond
	}

	// The user provided the right answer; succeed auth
	if e.Request.Session[l-1].ChallengeResult {
		response.IssueTokens = true
		response.FailAuthentication = false
		goto respond
	}

respond:
	e.Response = response
	return e, nil
}
```

**Comment by royjit** [comment]
Not an expert in GO, but looks like your lambda trigger always return `CUSTOM_CHALLENGE` as the response because of this line: `response := events.CognitoEventUserPoolsDefineAuthChallengeResponse{ChallengeName: "CUSTOM_CHALLENGE"}`

May if you change the last if statement it might work?:

```
if e.Request.Session[l-1].ChallengeResult {
		response.IssueTokens = true
		response.FailAuthentication = false
		response.ChallengeName = null // to make sure that no new challenge are send?
		goto respond
	}
```

**Comment by jranson** [comment]
@royjit thanks for your guidance on this, it is much appreciated. I've made this adjustment to their DefineAuthChallenge lambda clear the ChallengeName. The login flow was already fully working in a ReactJS web app, so fingers crossed this will get them going on the iOS side as well. @kevin-agrology will be testing shortly and we'll let you know!

**Comment by kevin-agrology** [comment]
Thanks @royjit but unfortunately we have the same result after the lambda trigger change. Can you think of any other potential source for errors on our end? Do you have a similar flow that you can test on your end?

**Comment by kevin-agrology** [comment]
@royjit We have it working now. Most likely there was a error in my amplify/aws config json files in the iOS client, which prompted @jranson to make some changes to make it work with the lambdas, but resulting in the confusing state of successful challenge but no authentication. Thanks so much for your help here!

**Comment by jranson** [comment]
@royjit Just to follow-up to Kevin's response. We have been following this blog entry to implement the custom auth lambdas: https://aws.amazon.com/blogs/mobile/implementing-passwordless-email-authentication-with-amazon-cognito/

The root cause for Agrology's issue was that the iOS Amplify library always sends an initial ChallengeName of SRP_A to the cognito signIn call.  However, the example "Define Auth Challenge trigger" from the blog ^^ is explicitly coded to fail any authentication calls where the ChallengeName is not CUSTOM_CHALLENGE. So we ported that same behavior over to our go-based version of these lambdas. Because the Define lambda looks for the CUSTOM_CHALLENGE ChallengeName and fails requests that have a different ChallngeName, the logic is incompatible with the iOS Amplify libraries as-is, since they initially send SRP_A.

I was able to work around this by modifying the Define Auth Challenge lambda to respond with the CUSTOM_CHALLENGE name instead of failing outright, and that seems to have fixed up the iOS side.
